### ISSUE TITLE
DispatchException

### ISSUE THREAD

**Issue by bernardnormier** [icerpc/icerpc-csharp#809]
# Proposal

1. Replace exceptions ServiceNotFoundException, OperationNotFoundException, UnhandledException and DispatchException by a single exception, DispatchException:

```
module Slice;

exception DispatchException
{
    errorCode: DispatchErrorCode,
}

unchecked enum DispatchErrorCode // unchecked as we may want to add more values over time
{
    ServiceNotFound,                 // could not find target
    OperationNotFound,            // target does not provide operation
    InvalidData,                          // could not decode header or payload (new)
    UnhandledException,           // dispatch failed with an unhandled exception
}
```

2. With Slice encoding 2, DispatchException is encoded/decoded like any other Slice remote exception.

3. With Slice encoding 1.1, DispatchException is encoded as one of the special "system exceptions", namely:

```
ServiceNotFound => ObjectNotExistException (with origin but no message)
OperationNotFound => OperationNotExistException (with origin but not message)
_ => UnknownException (with message but no origin). The encoded message starts with `[num] ` where num is the dispatch error code value, e.g. `[3] ` (see below).
```

For decoding, we map all 1.1 system exceptions to a DispatchException as follows:
```
ObjectNotExistException || FacetNotExistException => DispatchException(ServiceNotFound) with generic message and origin
OperationNotExistException => DispatchException(OperationNotFound) with generic message and origin
UnknownLocalException => DispatchException(UnhandledException) with message and empty origin
UnknownUserException =>  DispatchException(InvalidData) with message and empty origin
UnknownException => DispatchException(num) - see below
```

When we receive an UnknownException, we check if its message starts with `[num] ` (where num is a numeric value) and parse num into a DispatchErrorCode value. Then we create a DispatchException(num) with the message less `[num] ` and an empty origin.

Basically, we use `UnknownException`'s message to encode all DispatchException except ServiceNotFound and OperationNotFound.
 
If the message does not start with `[num] `, we decode the UnknownException into a DispatchException(UnhandledException).

One nice feature of this proposal is a ZeroC Ice client will get these new DispatchException as UnknownException with a somewhat usable message. For example, DispatchException(InvalidData) becomes something like:
```
UnknownException("[2] could not decode payload: exceeded max Slice decoding depth");
```
